app/services/user_service.rb

```diff
@@ -4,9 +4,27 @@ def self.delete_spam(user)
     user.destroy
   end
 
+  def self.deactivate(user:, actor:, params:)
+    actor.ability.authorize! :deactivate, user
+    UserDeactivationResponse.create(user: user, body: params[:deactivation_response])
+    user.deactivate!
+    user
+  end
+
   def self.update(user:, actor:, params:)
     actor.ability.authorize! :update, user
-    user.update! params
+    user.update params
+    user
+  end
+
+  def self.change_password(user:, actor:, params:)
+    actor.ability.authorize! :update, user
+    user.errors.add :current_password, I18n.t(:"error.current_password_did_not_match") unless user.valid_password? params.delete(:current_password)
```

Comment:
Yeah, I have change password covered with

- It doesn't work when the current password isn't provided correctly
- It doesn't work when the password doesn't match the password confirmation
- It works when both of those things are correct